
\SweaveOpts{engine=R, eps=FALSE, keep.source = TRUE}

<<warning, echo = FALSE, results = tex>>=
cat("\n\n\t%%%% DON'T EDIT THIS FILE\n\n")
@


<<setup, echo = FALSE, results = hide>>=
set.seed(290875)
options(prompt = "R> ")
options(SweaveHooks = list(mai = function() par(mai = par("mai") * c(1, 2, 1, 1))))
library("multcomp")
library("survival")
library("sandwich")
library("robustbase")
data("alpha", package = "coin")
data("bodyfat", package = "mboost")
data("alzheimer", package = "coin")
if (!file.exists("AML_Bullinger.rda"))
    load(url("http://www.imbe.med.uni-erlangen.de/~hothorn/AML_Bullinger.rda", open = "r"))
risk <- rep(0, nrow(clinical))
rlev <- levels(clinical[, "Cytogenetic.group"])
risk[clinical[, "Cytogenetic.group"] %in% rlev[c(7,8,4)]] <- "low"
risk[clinical[, "Cytogenetic.group"] %in% rlev[c(5, 9)]] <- "intermediate"
risk[clinical[, "Cytogenetic.group"] %in% rlev[-c(4,5, 7,8,9)]] <- "high"
risk <- as.factor(risk)
names(clinical)[6] <- "FLT3"
vcov.lmrob <- function(object) object$cov
@

\section{Illustrations} \label{illustrations}

\subsection{Genetic components of alcoholism} \label{alpha}

Various studies have linked alcohol dependence phenotypes to chromosome 4.
One candidate gene is \textit{NACP} (non-amyloid component of plaques),   
coding for alpha synuclein. 
\cite{Boenscheta2005} found longer alleles of
\textit{NACP}-REP1 in alcohol-dependent patients compared with healthy controls
and report that the allele lengths show some
association with levels of expressed alpha synuclein mRNA in
alcohol-dependent subjects (see Figure~\ref{alpha-box}). Allele length is
measured as a sum score built from additive dinucleotide repeat length and
categorized into three groups: short ($0-4$, $n = 24$), intermediate ($5-9$,
$n = 58$), and long ($10-12$, $n = 15$). Here, we are interested in comparing
the distribution of the expression level of alpha synuclein mRNA
in three groups of subjects defined by the allele length.

\setkeys{Gin}{width=0.7\textwidth}
\begin{figure}
\begin{center}
<<alpha-data-figure, echo = FALSE, fig = TRUE, width = 7, height = 5>>=
n <- table(alpha$alength)
par(cex.lab = 1.3, cex.axis = 1.3)
boxplot(elevel ~ alength, data = alpha, ylab = "Expression Level",
        xlab = "NACP-REP1 Allele Length", varwidth = TRUE)
axis(3, at = 1:3, labels = paste("n = ", n))
rankif <- function(data) trafo(data, numeric_trafo = rank)
@
\caption{\Robject{alpha} data: Distribution of levels of expressed alpha synuclein mRNA
         in three groups defined by the \textit{NACP}-REP1 allele lengths.
         \label{alpha-box}}
\end{center}
\end{figure}

Thus, we fit a simple one-way ANOVA model to the data and 
define $\K$ such that $\K \vartheta$ contains all three 
group differences:
<<alpha-aov-tukey, echo = TRUE>>=
data("alpha", package = "coin")
amod <- aov(elevel ~ alength, data = alpha)
amod_glht <- glht(amod, linfct = mcp(alength = "Tukey"))
amod_glht$linfct
@
The \Robject{amod\_glht} object now contains information
about the estimated linear function $\K \hat{\vartheta}$ and 
their covariance matrix $\hat{\S}$ which can be inspected
via the \Rcmd{coef()} and \Rcmd{vcov()} methods:
<<alpha-aov-coefvcov, echo = TRUE>>=
coef(amod_glht)
vcov(amod_glht)
@
The \Rcmd{summary()}
and \Rcmd{confint()} methods can be used to compute a
summary statistic including adjusted $p$-values and 
simultaneous confidence intervals, respectively:
<<alpha-aov-results, echo = TRUE>>=
confint(amod_glht)
summary(amod_glht)
@

Because of the variance heterogeneity 
that can be observed in Figure~\ref{alpha-box}, one might be concerned 
with the validity of the above results stating that there
is not difference between any combination of the three
allele lengths. A sandwich estimator $\hat{\Sigma}$ 
might be more appropriate in this situation, and
the \Rarg{vcov} argument of \Rcmd{glht()} can be used
to specify a function to compute some alternative 
estimator $\hat{\Sigma}$ as follows:
<<alpha-aov-tukey-sandwich, echo = TRUE>>=
amod_glht_sw <- glht(amod, linfct = mcp(alength = "Tukey"), 
                     vcov = sandwich)
summary(amod_glht_sw)
@
We used the \Rcmd{sandwich()} function from package \Rpackage{sandwich}
\citep{Zeileis2004, Zeileis2006} which provides us with a 
heteroskedasticity-consistent estimator of the covariance matrix.
This result is more in line with previously published findings
for this study
obtained from non-parametric test procedures such as the
Kruskal-Wallis test. A comparison of the simultaneous 
confidence intervals calculated based on the ordinary and
sandwich estimator is given in Figure~\ref{alpha-ci}.

\setkeys{Gin}{width=0.95\textwidth}
\begin{figure}
\begin{center}
<<alpha-confint-plot, echo = FALSE, fig = TRUE, width = 8, height = 3.5>>=
par(mai = par("mai") * c(1, 2.1, 1, 0.5))
layout(matrix(1:2, ncol = 2))
ci1 <- confint(glht(amod, linfct = mcp(alength = "Tukey")))
ci2 <- confint(glht(amod, linfct = mcp(alength = "Tukey"), vcov = sandwich))
plot(ci1, xlim = c(-0.6, 2.6), main = expression(paste("Tukey HSD (ordinary ", hat(Sigma), ")")), 
    xlab = "Difference", ylim = c(0.5, 3.5))
plot(ci2, xlim = c(-0.6, 2.6), main = expression(paste("Tukey HSD (sandwich ", hat(Sigma), ")")), 
    xlab = "Difference", ylim = c(0.5, 3.5))
@
\caption{\Robject{alpha} data: Simultaneous confidence intervals based
    on the ordinary covariance matrix (left) and a sandwich estimator (right). 
    \label{alpha-ci}}
\end{center}
\end{figure}


\subsection{Prediction of total body fat} \label{bodyfat}

\citet{garcia2005} report on the development of predictive regression equations
for body fat content by means of $p = 9$ common anthropometric
measurements which were obtained for $n = 71$ healthy German women.
In addition, the women's body composition was measured by
Dual Energy X-Ray Absorptiometry (DXA). This reference method
is very accurate in measuring body fat but finds little applicability
in practical environments, mainly because of high costs and the
methodological efforts needed. Therefore, a simple regression equation
for predicting DXA measurements of body fat is of special interest for the practitioner.
Backward-elimination was applied to select
important variables from the available anthropometrical measurements and
\citet{garcia2005} report a final linear model utilizing
hip circumference, knee breadth and a compound covariate which is defined as
the sum of log chin skinfold, log triceps skinfold and log subscapular skinfold.
Here, we fit the saturated model to the data and use the maximum-test
over all $t$-statistics to select important variables based on 
adjusted $p$-values. The linear model including all covariates and
the classical unadjusted $p$-values are given by
<<bodyfat-lm-fit, echo = TRUE>>=
data("bodyfat", package = "mboost")
summary(lmod <- lm(DEXfat ~ ., data = bodyfat))
@
The linear function $\K$ is basically the identity matrix, except 
for the intercept which is omitted. Once the matrix $\K$ has been 
defined, it can be used to set up the general linear hypotheses:
<<bodyfat-lm-maxtest, echo = TRUE>>=
K <- cbind(0, diag(length(coef(lmod)) - 1))
rownames(K) <- names(coef(lmod))[-1]
lmod_glht <- glht(lmod, linfct = K)
@
Classically, one would perform an $F$-test to check if
any of the regression coefficients is non-zero:
<<bodyfat-lm-Ftest, echo = TRUE>>=
summary(lmod_glht, test = Ftest())
@
but the source of the deviation from the global null-hypothesis
can only be inspected by the corresponding maximum-test, i.e., 
via
<<bodyfat-lm-maxtest, echo = TRUE>>=
summary(lmod_glht)
@
Only two covariates, waist and hip circumference, seem to be important
and caused the rejection of $H_0$. Alternatively, an MM-estimator
\citep{Yohai1987} as implemented by \Rcmd{lmrob()} from package 
\Rpackage{lmrob} \citep{pkg:robustbase} 
can be used to fit a robust version of the above linear
model, the results coincide rather nicely:
<<bodyfat-robust, echo = TRUE>>=
summary(glht(lmrob(DEXfat ~ ., data = bodyfat), linfct = K))
@

\subsection{Smoking and Alzheimer's disease} \label{alzheimer}

<<alzheimer-demographics, echo = FALSE>>=
total <- nrow(alzheimer)
stopifnot(total == 538) 
male <- sum(alzheimer$gender == "Male")
stopifnot(male == 200)
female <- sum(alzheimer$gender == "Female")
stopifnot(female == 338)
disease <- table(alzheimer$disease)
smoked <- sum(alzheimer$smoking != "None")
atab <- xtabs(~ smoking + + disease + gender, data = alzheimer)
### there is a discrepancy between Table 1 (32% smokers of 117 women
### suffering from other diagnoses) and Table 4 (63% non-smokers).  
### We used the data as given in Table 4.
@

\cite{SalibHillier1997}
report results of a case-control study on Alzheimer's disease
and smoking behavior of $\Sexpr{disease["Alzheimer's"]}$ female
and male Alzheimer patients and 
$\Sexpr{disease[names(disease) != "Alzheimer's"]}$ controls.
The \Robject{alzheimer} data 
%% shown in Table~\ref{alzheimertab}
have been 
re-constructed from Table~4 in \cite{SalibHillier1997}.
%% and are depicted in Figure~\ref{alz-plot}.
The authors conclude that `cigarette smoking is less frequent in
men with Alzheimer's disease.'
Originally, one was interested to assess whether there is any association
between smoking and Alzheimer's (or other dementia) diseases. Here, we 
focus on how a potential association can be described \citep[see][for
a non-parametric approach]{Hothorn:2006:AmStat}. 

First, we fit a logistic regression model including both main effects
and an interaction effect of smoking and gender. The response
is a binary variable giving the diagnosis of the patient (either suffering
from Alzheimer's disease or other dementias):
<<alzheimer-glm, echo = TRUE>>=
data("alzheimer", package = "coin")
y <- factor(alzheimer$disease == "Alzheimer's", 
    labels = c("other", "Alzheimer"))
gmod <- glm(y ~ smoking * gender, data = alzheimer, family = binomial())
summary(gmod)
@
The negative regression coefficient for heavy smoking males indicates
that Alzheimer's disease might be less frequent in this group, but the model
is still difficult to interpret based on the coefficients and 
corresponding $p$-values only. Therefore, confidence intervals
on the probability scale for the different `risk groups' are interesting 
and can be computed as follows. For each combination of gender and 
smoking behavior, the probability of suffering from Alzheimer's
disease can be estimated by computing the logit function 
of the linear predictor from model \Robject{gmod}. Using
the \Rcmd{predict()} method for generalized linear models
is a convenient way to compute these probability estimates. Alternatively,
we can set up a linear function $\K$ such that $\left(1 + \exp(- \K \hat{\vartheta})\right)^{-1}$
is the vector of estimated probabilities with simultaneous confidence intervals
\begin{eqnarray*}
\left(\left(1 + \exp\left(- \left(\K \hat{\vartheta} - q_\alpha \text{diag}\left(\hat{\S}^{-1/2}\right)\right)\right) \right)^{-1},
\left(1 + \exp\left(- \left(\K \hat{\vartheta} + q_\alpha \text{diag}\left(\hat{\S}^{-1/2}\right)\right)\right)\right)^{-1}\right)
\end{eqnarray*}

<<alzheimer-K, echo = FALSE>>=  
a <- cbind(levels(alzheimer$smoking), "Female")
b <- cbind(levels(alzheimer$smoking), "Male")
d <- rbind(a, b)
### colnames(d) <- c("smoking", "gender")
colnames(d) <- c("s", "g")
rownames(d) <- paste(d[,1], d[,2], sep = ":")
d <- as.data.frame(d)
K <- model.matrix(~ s * g, data = d)
colnames(K)[1] <- "(Icpt)"
attr(K, "assign") <- NULL
attr(K, "contrasts") <- NULL
@

For our model, $\K$ is given by the following matrix (essentially
the design matrix of \Robject{gmod} for eight persons with different
smoking behavior from both genders)
\small
<<alzheimer-K, echo = TRUE>>=
K
@
\normalsize
and can easily be used to compute the confidence intervals described above
<<alzheimer-probci, echo = TRUE>>=
gmod_ci <- confint(glht(gmod, linfct = K))
gmod_ci$confint <- apply(gmod_ci$confint, 2, binomial()$linkinv)
plot(gmod_ci, xlab = "Probability of Developing Alzheimer", 
     xlim = c(0, 1))
@
The simultaneous confidence intervals are depicted in 
Figure~\ref{alzheimer-plot}. Using this representation of the results,
it is obvious that Alzheimer's disease is less frequent in non-smoking and
heavy smoking men compared to all other configurations of the two covariates.

\begin{figure}
\begin{center}
<<alzheimer-plot, echo = FALSE, fig = TRUE, mai = TRUE, width = 7, height = 5>>=
plot(gmod_ci, xlab = "Probability of Developing Alzheimer", 
     xlim = c(0, 1), main = "", ylim = c(0.5, 8.5))
@
\caption{\Robject{alzheimer} data: Simultaneous confidence intervals for 
the probability to suffer from Alzheimer's disease. \label{alzheimer-plot}}
\end{center}
\end{figure}

\subsection{Acute myeloid leukemia} \label{AML}

The treatment of patients suffering from acute myeloid leukemia (AML) is
determined by a tumor classification scheme taking the status of various
cytogenetic aberrations into account. \cite{Bullingeretal2004} investigate an
extended tumor classification scheme incorporating molecular subgroups of the disease
obtained by gene expression profiling. The analyses reported here
are based on clinical data only (thus omitting available gene expression data) 
published online at 
\url{http://www.ncbi.nlm.nih.gov/geo}, accession number GSE425. The
overall survival time and censoring indicator as well as the   
clinical variables age, sex, lactic dehydrogenase level (LDH), 
white blood cell count (WBC), and treatment group are taken from Supplementary
Table 1 in \cite{Bullingeretal2004}. In
addition, this table provides two molecular markers, the fms-like tyrosine kinase 3 (FLT3) and
the mixed-lineage leukemia (MLL) gene, as well as 
cytogenetic information helpful to define a risk score (`low': karyotype
t(8;21), t(15;17) and inv(16); `intermediate': normal karyotype and t(9;11);   
and `high': all other forms). One interesting question might be the
usefulness of this risk score. Here, we fit a Weibull survival model
to the data including all above mentioned covariates as well
as their interactions with the patient's gender. Tukey's all-pair comparisons
highlight that there seems to be a difference between `high' scores
and both `low' and `intermediate' ones but the latter two aren't distinguishable:
<<bullinger-survreg, echo = TRUE>>=
smod <- survreg(Surv(time, event) ~ Sex * (Age + WBC + LDH + FLT3 + risk), 
                data = clinical)
summary(glht(smod, linfct = mcp(risk = "Tukey")))
@
